From 29e6895bd0d981cb82350a3226dcaa65c8be6682 Mon Sep 17 00:00:00 2001
From: Ryan Liptak <squeek502@hotmail.com>
Date: Sat, 8 Nov 2025 23:45:31 -0800
Subject: [PATCH 049/133] incr-check: Kill child process on error
X-Developer-Signature: v=1; a=openpgp-sha256; l=1588; i=xtex@astrafall.org;
 h=from:subject; bh=1xG7ThpMUTkcckjHjmw4p/eM5dpPUC07T5Pd20e7VqE=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78hhKexLalY023mpb9nTpr5m3WQ2tX3kEzXkr3f05I
 fnTOS3RjoksDGJcDJZiiixFhg3erDrp/KLLymVh5rAygQyRFmlgYGBgYGHgy03MKzXSMdIz1TbU
 MzTUMdIxYuDiFICpVrjC8E+bJ2Ut94LpG5ekvORsW/i+eLslw6UpD9YesJNj/pEdfnE7I8MeudN
 nXDLPyadt454otP3O4bmK/bWO5xZu3SguZ2EclccGAA==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

Since the child process is spawned with the tmp directory as its CWD, the child process opens it without DELETE access. On error, the child process would still be alive while the tmp directory is attempting to be deleted, so it would fail with `.SHARING_VIOLATION => return error.FileBusy`.

Fixes arguably the least important part of #22510, since it's only the directory itself that would fail to get deleted, all the files inside would get deleted just fine.

(cherry picked from commit 5358af7ba4cf16bef1734729646b99842e439916)
---
 tools/incr-check.zig | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/tools/incr-check.zig b/tools/incr-check.zig
index c187c84ae51e..90958df44c3d 100644
--- a/tools/incr-check.zig
+++ b/tools/incr-check.zig
@@ -185,6 +185,9 @@ pub fn main() !void {
         };
 
         try child.spawn();
+        errdefer {
+            _ = child.kill() catch {};
+        }
 
         var poller = std.Io.poll(arena, Eval.StreamEnum, .{
             .stdout = child.stdout.?,
@@ -591,6 +594,8 @@ const Eval = struct {
     fn fatal(eval: *Eval, comptime fmt: []const u8, args: anytype) noreturn {
         eval.tmp_dir.close();
         if (!eval.preserve_tmp_on_fatal) {
+            // Kill the child since it holds an open handle to its CWD which is the tmp dir path
+            _ = eval.child.kill() catch {};
             std.fs.cwd().deleteTree(eval.tmp_dir_path) catch |err| {
                 std.log.warn("failed to delete tree '{s}': {s}", .{ eval.tmp_dir_path, @errorName(err) });
             };
-- 
2.52.0

